File: Stacks_Queues/python/stack_of_plates.py
from LimitedStack import LimitedStack
from exceptions import SetEmptyException, WrongStackNumException


class SetOfStacks(object):
    def __init__(self, max_size):
        self.max_stack_size = max_size
        self.stacks = []
        self.stacks_num = 0

    def push(self, x):
        if not self.stacks_num:
            stack = LimitedStack(self.max_stack_size)
            self.stacks.append(stack)
            self.stacks_num += 1
        else:
            stack = self.stacks[self.stacks_num - 1]

        if stack.is_full():
            stack = LimitedStack(self.max_stack_size)
            self.stacks.append(stack)
            self.stacks_num += 1

        self.stacks[self.stacks_num - 1].push(x)

    def pop(self):
        return self.pop_at(self.stacks_num - 1)


    def pop_at(self, stack_num):
        if not self.stacks_num:
            raise SetEmptyException()

        if not 0 <= stack_num < self.stacks_num:
            raise WrongStackNumException(stack_num)

        stack = self.stacks[stack_num]
        top = stack.pop()

        if stack.is_empty():
            if stack_num == self.stacks_num - 1:
                self.stacks.pop()
            else:
                self.stacks = self.stacks[:stack_num] + self.stacks[stack_num+1:]
            self.stacks_num -= 1

        return top

    def print_set_of_stacks(self):
        for stack in self.stacks:
            print(stack.elements)
        print('\n')


s = SetOfStacks(2)
s.push(2)
s.print_set_of_stacks()
s.push(4)
s.push(1)
s.push(0)
s.push(1)
s.push(10)
s.print_set_of_stacks()

print(s.pop_at(1))
print(s.pop_at(1))
s.print_set_of_stacks()
print(s.pop_at(0))
print(s.pop_at(0))
s.print_set_of_stacks()
print(s.pop())
s.print_set_of_stacks()

File: Stacks_Queues/go/stack_of_plates.go
package main
import (
  "types"
  "fmt"
)


type SetOfStacks struct {
  MaxStackSize int
  stacks []*types.LimitedStack
  stacksNum int
}


func (set *SetOfStacks) Push(x int) {
  var stack *types.LimitedStack
  if set.stacksNum == 0 {
    stack = &types.LimitedStack{MaxSize: set.MaxStackSize}
    set.stacks = append(set.stacks, stack)
    set.stacksNum++
  } else {
    stack = set.stacks[set.stacksNum - 1]
  }

  if stack.IsFull() {
    stack := types.LimitedStack{MaxSize: set.MaxStackSize}
    stack.Push(x)
    set.stacks = append(set.stacks, &stack)
    set.stacksNum++
  } else {
    stack.Push(x)
  }
}


func (set *SetOfStacks) Pop() (int, bool) {
  return set.PopAt(set.stacksNum - 1)
}


func (set *SetOfStacks) PopAt(stackNum int) (int, bool) {
  var stack *types.LimitedStack
  if set.stacksNum == 0 {
    return 0, false
  }

  if stackNum < 0 || stackNum >= set.stacksNum {
    return 0, false
  }

  stack = set.stacks[stackNum]
  top, _ := stack.Pop()

  if stack.IsEmpty() {
    if stackNum == set.stacksNum - 1 {
      set.stacks = set.stacks[:set.stacksNum-1]
    } else {
      set.stacks = append(set.stacks[:stackNum], set.stacks[stackNum + 1:]...)
    }
    set.stacksNum--
  }

  return top, true
}

func (set *SetOfStacks) PrintSetOfStacks() {
  for i := 0; i < set.stacksNum; i++ {
    types.PrintLimitedStack(*set.stacks[i])
  }
  fmt.Println("\n")
}


func main() {
  s := SetOfStacks{MaxStackSize: 2}

  s.Push(3)
  s.Push(2)
  s.Push(1)
  s.Push(2)
  s.PrintSetOfStacks()
  s.Push(3)
  s.Push(4)
  s.PrintSetOfStacks()
  fmt.Println(s.PopAt(1))
  fmt.Println(s.PopAt(1))
  s.PrintSetOfStacks()
}
